arch/i386/mcount-support.c

```diff
@@ -0,0 +1,354 @@
+		0xff, 0x25, 0x00, 0x00, 0x00, 0x00,  /* jmp *(offset) */
+		0xcc, 0xcc, 0xcc, 0xcc, 0xcc,        /* padding */
+	};
+	Elf_Scn *sec;
+	const char *skip_syms[] = {
+		"mcount", "__fentry__",
+		"__cyg_profile_func_enter", "__cyg_profile_func_exit",
+		"__cxa_finalize",  /* XXX: it caused segfault */
+		"__gmon_start__",  /* XXX: it makes process stuck */
+	};
+	size_t shstr_idx;
+
+	dsymtab = &pd->dsymtab;
+
+	if (elf_getshdrstrndx(elf, &shstr_idx) < 0)
+		return -1;
+
+	sec = NULL;
+	while ((sec = elf_nextscn(elf, sec)) != NULL) {
+		GElf_Shdr shdr;
+		char *shname;
+
+		if (gelf_getshdr(sec, &shdr) == NULL)
+			return -1;
+
+		shname = elf_strptr(elf, shstr_idx, shdr.sh_name);
+		if (!strcmp(shname, ".plt"))
+			plt_addr = shdr.sh_addr + pd->base_addr;
+		if (!strcmp(shname, ".rela.plt"))
+			has_rela_plt = true;
+	}
+
+	if (plt_addr == 0) {
+		pr_dbg("cannot find PLT address\n");
+		return -1;
+	}
+
+	if (has_rela_plt) {
+		/* it's already handled by restore_plt_functions() in find_got() */
+		return 0;
+	}
+
+	trampoline_size = (dsymtab->nr_sym + 1) * sizeof(trampoline);
+	trampoline_buf = mmap(0, trampoline_size, PROT_READ | PROT_WRITE,
+			      MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
+	if (trampoline_buf == MAP_FAILED)
+		pr_err("failed to mmap trampoline for bind now");
+
+	pr_dbg2("setup bind-now PLT trampoline at %#lx\n", trampoline_buf);
+
+	for (idx = 0; idx < dsymtab->nr_sym; idx++) {
+		sym = dsymtab->sym_names[idx];
+
+		for (i = 0; i < ARRAY_SIZE(skip_syms); i++) {
+			if (!strcmp(sym->name, skip_syms[i]))
+				break;
+		}
+		if (i != ARRAY_SIZE(skip_syms))
+			continue;
+
+		memcpy(&r_offset, (void *)sym->addr + R_OFFSET_POS, sizeof(r_offset));
+		r_addr = r_offset + sym->addr + JMP_INSN_SIZE;
+
+		/* GOT index can be different than the symbol index */
+		got_idx = (r_addr - pltgot_addr) / sizeof(long);
+
+		target_addr = trampoline_buf + (idx * sizeof(trampoline));
+		real_addr = setup_pltgot(pd, got_idx, idx, target_addr);
+
+		jump_offset = (dsymtab->nr_sym - idx - 1) * sizeof(trampoline) + PAD_SIZE;
+
+		pr_dbg3("[%d] %s got idx %d, r_offset %lx, real address = %#lx, "
+			"target addr = %p, jump offset = %#lx\n",
+			idx, sym->name, got_idx, r_offset + sym->addr - pd->base_addr + JMP_INSN_SIZE,
+			real_addr, target_addr, jump_offset);
+
+		/* make up the instruction and copy to the trampoline buffer */
+		memcpy(&trampoline[PUSH_IDX_POS], &idx, sizeof(idx));
+		memcpy(&trampoline[JMP_OFS_POS], &jump_offset, sizeof(jump_offset));
+		memcpy(target_addr, trampoline, sizeof(trampoline));
+	}
+
+	pr_dbg2("real address to jump: %#lx\n", plt_addr);
+	memcpy(trampoline_buf + (idx * sizeof(trampoline)),
+	       &plt_addr, sizeof(plt_addr));
+
+	mprotect(trampoline_buf, trampoline_size, PROT_READ | PROT_EXEC);
+	return 0;
+}
+
+unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
+{
+	struct sym *sym;
+
+	sym = &pd->dsymtab.sym[idx];
+	return sym->addr + ARCH_PLTHOOK_ADDR_OFFSET;
+}
```

Comment:
If this is same as the default implementation, you can just drop it.

Reply:
default implement is 
```
libmcount/plthook.c:671:__weak unsigned long mcount_arch_plthook_addr(struct plthook_data *pd, int idx)
libmcount/plthook.c-672-{
libmcount/plthook.c-673-        struct sym *sym;
libmcount/plthook.c-674-
libmcount/plthook.c-675-        sym = &pd->dsymtab.sym[idx];
libmcount/plthook.c-676-        return sym->addr;
libmcount/plthook.c-677-}
libmcount/plthook.c-678-
```
ARCH_PLTHOOK_ADDR_OFFSET must be addition. 
I will upload another issue about this.

Reply:
You need to rebase your branch onto the current master.  There were some changes in this area.

Reply:
Now isn't it same? :)